Milestone 11 : Final Project Review

Nemwel-Boniface opened the pull request:
## Hello Reviewer(s) :wave: :wave: 
As I had made sure that I did most of the PR comments as details in all my previous PRs, I will do a brief summary of them all here:

In the Zoo project I was able to:

- [x] Set up the project including all the required libraries for the project
- [x] Choose an API that would give me numerical data to use
- [x] Set up Redux and my store
- [x] Connect my React components with my Reux store
- [x] Display 10 random animals on each page refresh. Note that the project does not re-render on the switch of the routes
- [x] Create the Details page and made sure that I am displaying the information of the right animal
- [x] Mae tests for both my components as well as the Redux store
- [x] Document my project professionally by making sure that I have a professional README.md file
- [x] Check to make sure that I met all the project requirements
- [x] For this project made sure  to use correct git-flow practices
- [x] On each milestone, I made sure to check and fix all the linter errors


Thank you for taking the time to review my project :pray:

Nemwel-Boniface committed:
Update:
 - Renamed APP components to match the project
 - Re orhanized the projects file structure making it more organized

Nemwel-Boniface committed:
Update:
 - Update included font awesome icons cdn links
 - Update included the lato font link
 - Updated the project tit;e to match the project

Nemwel-Boniface committed:
Update:
 - Included the Redux folder put as separate for better file structure
 - Included the ZooConfig which is to contain the project store
 - Included Zoo.js which will contain the project reducer and action creators
 - Included the actionTypes.js which will include all the action types for the project

Nemwel-Boniface committed:
Update removed bolier plate data and included the zoo component

Nemwel-Boniface committed:
Included a starting placeholder text for testing purposes

Nemwel-Boniface committed:
Update checked and fixed all linter errors

Nemwel-Boniface committed:
Merge pull request #1 from Nemwel-Boniface/setUp

Milestone 1 : Project Set up

Nemwel-Boniface committed:
Update included header component markup

Nemwel-Boniface committed:
Update renamed stylesheets file names and included stylesheets to project

Nemwel-Boniface committed:
Update included basic styling for the header/ navbar section of project

Nemwel-Boniface committed:
Renamed files

Nemwel-Boniface committed:
Update:
 - Included global styling for project giving it a project font family
 - Included some padding to the navbar to make it more visually apealling

Nemwel-Boniface committed:
Update checked and fixed linter errors

Nemwel-Boniface committed:
Merge pull request #2 from Nemwel-Boniface/header

Milestone 2 : project Header

Nemwel-Boniface committed:
Update Included:
 - Import for the get animals from API action type
 - Zooanimals reducer setup
 - Link to the API
 - Setup for the fetch animals from API action creator

Nemwel-Boniface committed:
Update included the fetch from API action and exported it

Nemwel-Boniface committed:
Update included:
 - Set up the Fetch animals action type to fetch the random animals from the API
 - Set the reducer to handle all the returned animals from the API

Nemwel-Boniface committed:
Update to the project action types to export the action as a default

Nemwel-Boniface committed:
Update checked and fixed redux reducer linter errors

Nemwel-Boniface committed:
Merge pull request #3 from Nemwel-Boniface/ReduxSetup

Milestone 3 : Redux setup

Nemwel-Boniface committed:
Update included redux logger lib to project

Nemwel-Boniface committed:
Update:
 - Imported all necessary libs to be used by my store
 - Created my store and exported it as a default
 - Connected the store with redux logger to help me in debugging to console the action dispatched

Nemwel-Boniface committed:
Update checked and fixed all store linter errors

Nemwel-Boniface committed:
Merge pull request #4 from Nemwel-Boniface/StoreSetUp

Milestone 4 : Store set up

Nemwel-Boniface committed:
Update included the combine reducer

Nemwel-Boniface committed:
Update to the Zoo component file

Nemwel-Boniface committed:
Update to the individual animal component

Nemwel-Boniface committed:
Update included proptypes dep

Nemwel-Boniface committed:
Update did props validation

Nemwel-Boniface committed:
Update checked and fixed linter errors

Nemwel-Boniface committed:
Merge pull request #5 from Nemwel-Boniface/diplayAnimals

Milestone 5 : Display animals

Nemwel-Boniface committed:
Update styled the animals placing them in 2 columns

Nemwel-Boniface committed:
Update included Lnanding component to project

Nemwel-Boniface committed:
Update included a landing page that features a lion with some text

Nemwel-Boniface committed:
Update styled the landing component

Nemwel-Boniface committed:
Update included the landing image

Nemwel-Boniface committed:
Update cecked and fixed all linter errors

Nemwel-Boniface committed:
Merge pull request #6 from Nemwel-Boniface/animalsStyling

Milestone 6 : Animals styling & Landing Section

Nemwel-Boniface committed:
Updated the marku to have a link to see more details about an animal

Nemwel-Boniface committed:
Update to the Animal component styling to make it match the provided design

Nemwel-Boniface committed:
Update made general project styling eg changing text color to white

Nemwel-Boniface committed:
Update created the details component with placeholder text

Nemwel-Boniface committed:
Update setup project routes

Nemwel-Boniface committed:
Update included the Browser router in the project

Nemwel-Boniface committed:
Update refactored to include dispatch function in main component

Nemwel-Boniface committed:
Update setup the details component which displays correctly

Nemwel-Boniface committed:
Update included all the placeholder text for each inividual animal

Nemwel-Boniface committed:
Update styled the details component

Nemwel-Boniface committed:
Update checked and fixed alllinter errors

Nemwel-Boniface committed:
Merge pull request #7 from Nemwel-Boniface/dynamicRouting

Milestone 7 : Dynamic routing

Nemwel-Boniface committed:
Update made landing page to only show on the main page alone

Nemwel-Boniface committed:
Update checked and fixed milesotone linter errors

Nemwel-Boniface committed:
Merge pull request #8 from Nemwel-Boniface/hideHeader

Milestone 8 : Hide Lion's landing page

Nemwel-Boniface committed:
Update installed react test renderer

Nemwel-Boniface committed:
Update included a test for the Home page

Nemwel-Boniface committed:
Update included test for each individual animal

Nemwel-Boniface committed:
Update test for the initial state of the reducer

Nemwel-Boniface committed:
Update checked whether the number of items returned is an array of 10 emements

Nemwel-Boniface committed:
Update check whether the reducer returns a different item

Nemwel-Boniface committed:
Update checked and fixed miestone linter errors

Nemwel-Boniface committed:
Merge pull request #9 from Nemwel-Boniface/Tests

Milestone 9 : Tests

Nemwel-Boniface committed:
Update to README.md file as first dratf of project decumentation

Nemwel-Boniface committed:
Update included screenshot images for README.md file

Nemwel-Boniface committed:
Merge pull request #10 from Nemwel-Boniface/projectDocumentation

Milestone 10 : Project documentation

TedLivist commented on src/components/details.js:
```diff
@@ -0,0 +1,77 @@
+    });
+  }, []);
+  return (
+    <div className="Details">
+      <button type="button" onClick={() => push('/')}>
+        <i className="fa fa-arrow-left" aria-hidden="true" />
+        Home
+      </button>
```
- [ ] As indicated in the design the **`Back`** button (<) should be placed in the design, the back arrow for the **Details** page should be on the header. Kindly fix that.

  ![nemwel-back-edit](https://user-images.githubusercontent.com/36542986/164275400-6e14de8d-f9f7-4204-8eeb-8a059e0e7eb7.jpg)

Nemwel-Boniface replied:
Hello @TedLivist thank you for this review.
I have effected this change

TedLivist commented on src/components/Zoo.js:
```diff
@@ -0,0 +1,38 @@
+import Animal from './Animal';
+import Landing from './landing';
+
+const Zoo = () => {
+  const animals = useSelector((state) => state.animals, shallowEqual);
+
+  return (
+    <div className="Zoo">
+      <Landing />
+      <ul className="zooanimals">
+        {animals.map((animal) => (
+          <Animal
+            key={animal.id}
+            animal={animal}
+            name={animal.name}
+            latinName={animal.latin_name}
+            animalType={animal.animal_type}
+            activeTime={animal.active_time}
+            minLength={animal.length_min}
+            maxLength={animal.length_max}
+            minWeight={animal.weight_min}
+            maxWeight={animal.weight_max}
+            lifeSpan={animal.lifespan}
+            habitats={animal.habitat}
+            diet={animal.diet}
+            locattions={animal.geo_range}
+            image={animal.image_link}
+          />
+
+        ))}
```
- [ ] According to the design we have alternating lighter and darker colors for the contents on the homepage. Kindly implement that.

  Your Implementation | Given Design
  :---:|:---:
  ![nemwel-alternating](https://user-images.githubusercontent.com/36542986/164277701-7fea9dc1-0edd-4eb3-b900-8e73271f5225.jpg) | ![jonathas-orig-home](https://user-images.githubusercontent.com/36542986/164277484-3ae53deb-c2e0-43e4-b205-0324964deb2e.jpg)

TedLivist commented on src/components/details.js:
```diff
@@ -0,0 +1,77 @@
+          Animal Type
+          <span className="right">{animal.animalType}</span>
+        </p>
+        <p>
+          Active Times
+          <span className="right">{animal.activeTime}</span>
+        </p>
+        <p>
+          Minimum Length
+          <span className="right">{animal.minLength}</span>
+        </p>
+        <p>
+          Maximum Length
+          <span className="right">{animal.maxLength}</span>
+        </p>
+        <p>
+          Minimum Weight
+          <span className="right">{animal.minWeight}</span>
+        </p>
+        <p>
+          Maximum Weight
+          <span className="right">{animal.maxWeight}</span>
+        </p>
+        <p>
+          Life span
+          <span className="right">{animal.lifeSpan}</span>
+        </p>
+        <p>
+          Habitats
+          <span className="right">{animal.habitats}</span>
+        </p>
+        <p>
+          Diet
+          <span className="right">{animal.diet}</span>
+        </p>
+        <p>
+          Locations
+          <span className="right">{animal.locattions}</span>
+        </p>
+      </div>
```
- [ ] Let's try to follow the design as much as possible. On your Details page, kindly add the stripe color as we have in the template.

- [ ] Ensure you maintain the appropriate spacing between elements.

  Your Implementation | Behance Design
  :---:|:---:
  ![nemwel-details](https://user-images.githubusercontent.com/36542986/164278614-97edebb9-32f8-4d5b-a3dd-d3cd6ee2743f.jpg) | ![jonathas-origin-detail](https://user-images.githubusercontent.com/36542986/164278607-101d01d3-0937-480d-8629-6e5fd891e678.jpg)

Nemwel-Boniface replied:
Hello @TedLivist Thank you I have effected the change.

TedLivist commented on src/components/landing.js:
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import lnd from '../images/Lion.png';
+
+const Landing = () => (
+  <div className="Landing">
+    <img src={lnd} alt="Gif" />
+    <div className="LandingDetails">
+      <h2>Lions</h2>
+      <p>The king of the jungle</p>
+    </div>
+  </div>
+);
+export default Landing;
```
- [ ] Your landing section on the homepage is not as descriptive of the elements on the homepage as what we have on the design. In the design, **EUROPE** and a map of it describes the content which includes the Czech Republic, Italy and so on which are typically countries in the European continent. In your implementation, **LION** and its image, does not describe the homepage contents. Kindly resolve this.

  Your Implementation | Given Design
  :---:|:---:
  ![nemwel-landing](https://user-images.githubusercontent.com/36542986/164276896-043c99d4-1b0f-45e6-942e-3f70033c7955.jpg) | ![akumu-descriptive](https://user-images.githubusercontent.com/36542986/164276575-6d6fce0f-cde9-43ab-97c8-3384ac5c1f74.jpg)

TedLivist commented on src/components/Zoo.js:
```diff
@@ -0,0 +1,38 @@
+import Animal from './Animal';
+import Landing from './landing';
+
+const Zoo = () => {
```
- [ ] On your homepage, kindly ensure you implement the filtering mechanism (like a search box) as outlined in the requirements.

  ![furkan-home](https://user-images.githubusercontent.com/36542986/164279679-888ccfe5-7a9d-4dd9-b303-f9f05efcbd35.jpg)

Nemwel-Boniface replied:
Hello @TedLivist I have effected this requested additional feature

TedLivist commented on src/App.test.js:
```diff
@@ -1,8 +1,8 @@
 import { render, screen } from '@testing-library/react';
-import App from './App';
+import ZooContainer from './ZooContainer';
 
 test('renders learn react link', () => {
-  render(<App />);
+  render(<ZooContainer />);
   const linkElement = screen.getByText(/learn react/i);
   expect(linkElement).toBeInTheDocument();
 });
```
- [ ] One of your tests is failing. Kindly resolve it.

  ![nemwel-test](https://user-images.githubusercontent.com/36542986/164280493-f3cd528c-11e8-46e5-8026-e5755817ac9b.jpg)

TedLivist commented on README.md:
```diff
@@ -1,20 +1,43 @@
+![](https://img.shields.io/badge/Microverse-blueviolet)
 
-# React-Redux Setup
+# Nemwel ZOO
```
- [ ] Currently, you have only one unit test for the Zoo component. In your app, kindly add more unit tests for some of your components.

- [ ] **Kindly ensure all implemented changes are pushed to the live demo.**

TedLivist reviewed (changes requested):
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @Nemwel-Boniface,

Good job so far!
✔️ Good documentation
✔️ Connection to the store using Redux hooks
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Nemwel-Boniface committed:
Update made style changes requested by reviewer

Nemwel-Boniface committed:
Update included tests for the details component

Nemwel-Boniface committed:
Update included back link in navbar

Nemwel-Boniface committed:
Update implemented conditional rendering of background images

Nemwel-Boniface committed:
Update included the search to filter functionality

Nemwel-Boniface committed:
Update included:
 - The markup for alternating shades of color for the animals
 - The styling for the different shades of the background

Nemwel-Boniface committed:
Styled the home link to match the project's wireframe

Nemwel-Boniface committed:
Update to the landing page contents

Nemwel-Boniface committed:
Update styled the search bar in the landing page

Nemwel-Boniface committed:
Update included the landing page image of the big 6 animals

Nemwel-Boniface committed:
Update to the general projectmarkup

Nemwel-Boniface committed:
Update included more unit tests and fixed failing tests

Nemwel-Boniface committed:
Update checked and fixed al milestone linter errors

Nemwel-Boniface committed:
Updated the README.md file and images to reflect the new changes

Nemwel-Boniface committed:
Updated the README.md file to include the live demo

Meltrust reviewed (approved):
Hi @Nemwel-Boniface,


## Wow, you did it 🎉
## Congratulations on finishing React!!!
![Brilliant](https://media.giphy.com/media/d31w24psGYeekCZy/giphy.gif)
## Thank you for adding the back button, implementing the darker color, implement correct spacing  from the elements, making the landing section more descriptive, adding a filtering mechanism, and adding more tests, as requested by the previous reviewer 💪 🥇 ㊗️ 
### Well done!




Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
